---
title: "Code Review"
author: "Annika Ladwig"
output:
  prettydoc::html_pretty:
    theme: HPSTR
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Workflow

# 1 Fastq-dump (load ENCODE fastq files)
```{r, engine = 'bash', eval = FALSE}
fastq-dump --gzip --skip-technical --readids --split-3 --outdir  /share/project/zarnack/annika/01_fastq_files/endodermal_cell/ SRR16809044
fastq-dump --gzip --skip-technical --readids --split-3 --outdir  /share/project/zarnack/annika/01_fastq_files/endodermal_cell/ SRR16809045
fastq-dump --gzip --skip-technical --readids --split-3 --outdir  /share/project/zarnack/annika/01_fastq_files/endodermal_cell/ SRR16809043

fastq-dump --gzip --skip-technical --readids --split-3 --outdir  /share/project/zarnack/annika/01_fastq_files/H1_cell_line/ SRR16811699
fastq-dump --gzip --skip-technical --readids --split-3 --outdir  /share/project/zarnack/annika/01_fastq_files/H1_cell_line/ SRR16811701
fastq-dump --gzip --skip-technical --readids --split-3 --outdir  /share/project/zarnack/annika/01_fastq_files/H1_cell_line/ SRR16811700
```

# 2 Trimmomatic (read processing) --> test impact of read length
```{r, engine = 'bash', eval = FALSE}
export PATH=~/miniconda3/bin:$PATH

# endodermal cell
cd /share/project/zarnack/annika/01_fastq_files/endodermal_cell
read_length=(100 75 50 36)

# trim adapters, filter 100 nt long reads, trim reads to 100, 75, 50, 36 nt
for length in ${read_length[@]}; do
  for SRR in `ls -1 *_1.fastq.gz | sed 's/\_1.fastq.gz//'`; do 
    echo trimmomatic PE $SRR\_1.fastq.gz $SRR\_2.fastq.gz \
    $length\_$SRR\_R1_paired.fastq.gz $length\_$SRR\_R1_unpaired.fastq.gz $length\_$SRR\_R2_paired.fastq.gz $length\_$SRR\_R2_unpaired.fastq.gz \
    ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:8:true \
    SLIDINGWINDOW:4:20 \
    MINLEN:100 \
    CROP:$length >> /share/project/zarnack/annika/02_processed_reads/endodermal_cell/cmd_file_$length
  done
  bash /share/project/zarnack/annika/02_processed_reads/endodermal_cell/cmd_file_$length
done
```

# 3 FastQC (quality report)
```{r, engine = 'bash', eval = FALSE}
fastqc *.bam --o /share/project/zarnack/annika/03_star_mapping/U2AF2/100/
```

# 4 STAR
```{r, engine = 'bash', eval = FALSE}
export PATH=~/miniconda3/bin:$PATH
source activate

read_length=(100 75 50 36)
gtf=/share/project/zarnack/annika/gencode.v41.annotation.gtf
genome=/share/project/zarnack/annika/GRCh38.p13.genome.fa
gd=/share/project/zarnack/annika/genome_indices

# GENERATING GENOME INDEXES for all read length
for i in ${read_length[@]}; do
  STAR \
    --runThreadN 6 \
    --runMode genomeGenerate \
    --genomeDir $gd/$i \
    --genomeFastaFiles $genome \
    --sjdbGTFfile $gtf \
    --sjdbOverhang $i-1
done 

# MAPPING READS
for length in ${read_length[@]}; do
  for SRR in `ls -1 $length\_*_R1_paired.fastq.gz | sed 's/\_R1_paired.fastq.gz//'`; do  
    STAR \
      --runThreadN 6 \
      --genomeDir $gd/$length \
      --readFilesCommand zcat \
      --readFilesIn $SRR\_R1_paired.fastq.gz $SRR\_R2_paired.fastq.gz \
      --outFileNamePrefix /share/project/zarnack/annika/03_star_mapping/endodermal_cell/$SRR\_ \
      --outFilterMultimapNmax 1 \
      --outFilterMismatchNmax 999 \
      --outFilterMismatchNoverLmax 0.04 \
      --outSAMtype BAM SortedByCoordinate
    done
done
```

# 5 Samtools view (extract specific amount of reads) --> test impact of read depth
```{r, engine = 'bash', eval = FALSE}
# shrink read depth to 30000 25000000 20000000...1000000 reads

# CT 1 
# ~ 30000000 reads
samtools view -b -s 0.96323430963552176957701491760975332532564543923003 U2AF2_100_CT_1.bam > read_depth/U2AF2_CT_1_30M_reads.bam

# ~ 25000000 reads
samtools view -b -s 0.80269525802960147464751243134146110443803786602502 U2AF2_100_CT_1.bam > read_depth/U2AF2_CT_1_25M_reads.bam

# ~ 20000000 reads
samtools view -b -s 0.64215620642368117971800994507316888355043029282002 U2AF2_100_CT_1.bam > read_depth/U2AF2_CT_1_20M_reads.bam

# ~ 15000000 reads
samtools view -b -s 0.48161715481776088478850745880487666266282271961501 U2AF2_100_CT_1.bam > read_depth/U2AF2_CT_1_15M_reads.bam#

# ...
```

# 6 MAJIQ - LSV detection and quantification

## 6.1 MAJIQ build - LSV detection

### 6.1.1 - test read length

for each read length a config file (100, 75, 50, 36 nt)

```
[info]
readlen=100
bamdirs=/share/project/zarnack/annika/03_star_mapping/U2AF2/100/read_depth
genome=GRCh38
strandness=reverse

[experiments]
U2AF2_100_KD=U2AF2_100_KD_1,U2AF2_100_KD_2
U2AF2_100_CT=U2AF2_100_CT_1,U2AF2_100_CT_2
```

```{r, engine = 'bash', eval = FALSE}
source /home/mario/applications/MAJIQv2.3/bin/activate

# FILES AND DIRS
GFF3="/share/project/zarnack/annika/gencode.v41.annotation.gff3"
CONFIGFILE="/share/project/zarnack/annika/04_majiq/build"
OUTDIR="/share/project/zarnack/annika/04_majiq/build"

source /home/mario/applications/MAJIQv2.3/bin/activate

# 2 samples
for combi in ${combos_2er[@]}; do
  majiq build $GFF3 \
  -c $CONFIGFILE/SRSF6/100/2_samples/sample$combi/SRSF6_100_config_file_2_samples$combi.txt \
  -j 6 \
  -o $OUTDIR/SRSF6/100/2_samples/sample$combi
done
```

### 6.1.2 - test read depth
for each read depth a config file
here: bam files with 30M reads
[info]
readlen=100
bamdirs=/share/project/zarnack/annika/03_star_mapping/U2AF2/100/read_depth
genome=GRCh38
strandness=reverse

[experiments]
U2AF2_30M_reads_KD=U2AF2_KD_1_30M_reads,U2AF2_KD_2_30M_reads
U2AF2_30M_reads_CT=U2AF2_CT_1_30M_reads,U2AF2_CT_2_30M_reads

```{r, engine = 'bash', eval = FALSE}
reads=(30M 25M 20M 15M 10M 5M 2.5M 1M)

for reads in ${reads[@]}; do
  majiq build $GFF3 \
  -c $CONFIGFILE/U2AF2/read_depth/$reads/U2AF2_${reads}_reads_config_file.txt \
  -j 6 \
  -o $OUTDIR/U2AF2/read_depth/$reads
done
```

### 6.1.3 - test sample size

for each sample size combination a config file
2 vs 3 samples and 3 different combinations for the 2 samples
[info]
readlen=100
bamdirs=/share/project/zarnack/annika/03_star_mapping/endodermal_cell/,/share/project/zarnack/annika/03_star_mapping/H1_cell_line/
genome=GRCh38
strandness=reverse

[experiments]
endodermal_100_samples_1_2=endodermal_cell_100_R1,endodermal_cell_100_R2
H1_100_samples_1_2=H1_cell_line_100_R1,H1_cell_line_100_R2


```{r, engine = 'bash', eval = FALSE}
#FILES AND DIRS
GFF3="/share/project/zarnack/annika/gencode.v41.annotation.gff3"
BUILD="/share/project/zarnack/annika/04_majiq/build/endodermal_vs_H1/100"

# SAMPLE COMBINATIONS
combos_2er=('_1_2' '_1_3' '_2_3')

# MAJIQ BUILD 
# 2 samples
for combi in ${combos_2er[@]}; do
  majiq build $GFF3 \
  -c $BUILD/2_samples/sample$combi/endo_vs_H1_100_config_file_2_samples$combi.txt \
  -j 6 \
  -o $BUILD/2_samples/sample$combi
done

# 3 samples
majiq build $GFF3 \
-c $BUILD/3_samples/endo_vs_H1_100_config_file_3_samples.txt \
-j 6 \
-o $BUILD/3_samples
```

## 6.2 MAJIQ dpsi - LSV quantification
```{r, engine = 'bash', eval = FALSE}
# is the same for all tests
# do the same for all read length, for all read depth and sample combinations
source /home/mario/applications/MAJIQv2.3/bin/activate

#FILES AND DIRS
BUILDDIR="/share/project/zarnack/annika/04_majiq/build"
OUTDIR="/share/project/zarnack/annika/04_majiq/dpsi"

genes=(SRSF6)
read_length=(100)

# 4 samples
for gene in ${genes[@]}; do
  for length in ${read_length[@]}; do
  majiq deltapsi \
  -grp1 \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_KD_1.majiq" \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_KD_2.majiq" \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_KD_3.majiq" \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_KD_5.majiq" \
  -grp2 \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_CT_1.majiq" \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_CT_2.majiq" \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_CT_3.majiq" \
  "${BUILDDIR}/$gene/$length/4_samples/sample_1_2_3_5/${gene}_${length}_CT_5.majiq" \
  --names $gene\_$length\_KD_4_samples $gene\_$length\_CT_4_samples \
  -j 6 \
  --output-type voila \
  -o $OUTDIR/$gene/$length/4_samples/sample_1_2_3_5
  done
done
```

## 6.3 MAJIQ voila
```{r, engine = 'bash', eval = FALSE}
# is the same for all tests
# do the same for all read length, for all read depth and sample combinations
source /home/mario/applications/MAJIQv2.3/bin/activate

#FILES AND DIRS
BUILDDIR="/share/project/zarnack/annika/04_majiq/build/"
DPSIDIR="/share/project/zarnack/annika/04_majiq/dpsi/"
OUTDIR="/share/project/zarnack/annika/04_majiq/voila/"

genes=(U2AF2 UPF1)
readLength=(100 75 50 36)

# MAJIQ Voila
# U2AF2 and UPF1
for GENE in ${genes[@]}; do
  for LENGTH in ${readLength[@]}; do
    voila tsv \
    -f "${OUTDIR}/${GENE}/thresholds_0.05/${LENGTH}/${GENE}_${LENGTH}_KD_vs_CT_voila.tsv" \
    --show-all \
    --threshold 0.05 \
    --changing-between-group-dpsi 0.05 \
    --non-changing-between-group-dpsi 0.05 \
    -j 6 \
    "${BUILDDIR}/${GENE}/${LENGTH}/splicegraph.sql" \
    "${DPSIDIR}/${GENE}/${LENGTH}/${GENE}_${LENGTH}_KD-${GENE}_${LENGTH}_CT.deltapsi.voila"   
  done   
done
```

# 7 MAJIQ Modulize - LSV classification
```{r, engine = 'bash', eval = FALSE}
source /home/mario/applications/MAJIQv2.3/bin/activate

#FILES AND DIRS
BUILDDIR="/share/project/zarnack/annika/04_majiq/build/U2AF2/read_depth"
DPSIDIR="/share/project/zarnack/annika/04_majiq/dpsi/U2AF2/read_depth"
OUTDIR="/share/project/zarnack/annika/06_modulize/U2AF2/read_depth"

read_depth=(30M 25M 20M 15M 10M 5M 2_5M 1M)

# voila modulize U2AF2
for reads in ${read_depth[@]}; do
  voila modulize \
  --changing-between-group-dpsi 0.05 \
  --non-changing-between-group-dpsi 0.05 \
  --changing-between-group-dpsi-secondary 0.025 \
  --show-all \
  -j 6 \
  -d "${OUTDIR}/${reads}" \
  "${BUILDDIR}/${reads}/splicegraph.sql" \
  "${DPSIDIR}/${reads}/U2AF2_KD_${reads}_reads-U2AF2_CT_${reads}_reads.deltapsi.voila"
done
```



