---
title: "TPM analysis"
author: "Annika Ladwig"
output:
  prettydoc::html_pretty:
    theme: architect
---

```{r, echo=FALSE, warning=FALSE,message=FALSE,error=FALSE,fig.keep='all'}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# LIBRARIES
library(dplyr)
library(ggplot2)
library(stringr)
library(ggpubr)
library(GenomicRanges)
library(GenomicFeatures)
library(ggpointdensity)
library(viridis)
library(RColorBrewer)
library(colorspace)

# theme
source(".../theme_paper.R")
theme_set(theme_paper())

# HELPER FUNCTIONS
# function to count occurrences of genes in df and create new data frame with this info
count_gene_occurrences <- function(df) {
  df |>
    group_by(gene_name) |>
    summarise(count = n())
}

# function to extract max. dPSI for each LSV and add new column with this information
extractMaxDPSIs <- function(x) {
  maxDeltaPSIs = x$mean_dpsi_per_lsv_junction %>%
    str_split(pattern = ";") %>%
    sapply(FUN = function(dPSIs){
      return(
        dPSIs %>% as.numeric %>% abs %>% max
      )
    })
  x$maxDeltaPSI = maxDeltaPSIs
  x
}

# function to extract max. probability for each LSV and add new column with this information
extractMaxProbs <- function(x) {
  maxProbs = x$probability_changing %>%
    str_split(pattern=";") %>%
    sapply(FUN=function(probs){
      return(
        probs %>% as.numeric %>% max
      )
    })
  x$maxProb = maxProbs
  x
}

# function to load files as data frames
loadFiles <- function(x){
  x = read.table(file = x,
                 header = TRUE,
                 sep = "\t",
                 stringsAsFactors = FALSE)
}
```


# Load data

## Load htseq-count files

Two knockdown and two control replicate files of the U2AF2 knockdown experiment are loaded. In order to calculate the TPMs later, the mean of the raw gene counts of all four replicates is taken.The data set with the 100 nt long reads was taken. 

### For U2AF2 KD experiment

```{r}
# load htseq count files for U2AF2 KD experiment (read length 100 nt)
htseqDir_RL_100 = ".../htseq_count_files_U2AF2_KD"
htseqFiles_RL_100=list.files(path = htseqDir_RL_100, pattern = ".tabular")

# load gene count files as data frames
setwd(htseqDir_RL_100)
counts_RL_100 = lapply(htseqFiles_RL_100, loadFiles)

# add column names
add_header <- function(df, new_header) {
  colnames(df) <- NULL
  colnames(df) <- new_header
  return(df)
}

counts_RL_100 = lapply(counts_RL_100, add_header, c("gene_id", "count"))

# find common gene IDs in 4 replicates for each read lengths
common_gene_ids_RL_100 <- Reduce(intersect, lapply(counts_RL_100, `[[`, "gene_id"))

# create new data frame with common gene IDs
mean_counts_RL_100 <- data.frame(gene_id = common_gene_ids_RL_100)

# calculate mean counts for each gene id from all conditions (2 x kd, 2 x ct)
mean_counts_RL_100 <- mean_counts_RL_100 %>%
  left_join(counts_RL_100[[1]], by = "gene_id") %>%
  left_join(counts_RL_100[[2]], by = "gene_id") %>%
  left_join(counts_RL_100[[3]], by = "gene_id") %>%
  left_join(counts_RL_100[[4]], by = "gene_id") %>%
  mutate(mean_counts = rowMeans(dplyr::select(., starts_with("count")), na.rm = TRUE)) %>%
  dplyr::select(-2, -3, -4, -5)
```


### For H1 cell line und endodermal cells

Here all files (3 x endodermal, 3 x H1 cell line) are loaded together. There is no differentiation between 2 and 3 samples here. This only occurs when MAJIQ is used. 

```{r}
# load htseq count files for all combinations of replicates
htseqDir_SS = ".../htseq_count_files_H1_vs_endodermal"
htseqFiles_SS=list.files(path = htseqDir_SS, pattern = ".tabular")
htseqFiles_SS

# load gene count files as data frames
setwd(htseqDir_SS)
counts_SS = lapply(htseqFiles_SS, loadFiles)

# add column names
add_header <- function(df, new_header) {
  colnames(df) <- NULL
  colnames(df) <- new_header
  return(df)
}

counts_SS = lapply(counts_SS, add_header, c("gene_id", "count"))

# find common gene IDs
common_gene_ids_SS <- Reduce(intersect, lapply(counts_SS, `[[`, "gene_id"))

# create new data frame with common gene IDs
mean_counts_SS <- data.frame(gene_id = common_gene_ids_SS)

# calculate mean counts for each gene id from all conditions (3 x endodermal cell, 3 x H1 cell line)
mean_counts_SS <- mean_counts_SS %>%
  left_join(counts_SS[[1]], by = "gene_id") %>%
  left_join(counts_SS[[2]], by = "gene_id") %>%
  left_join(counts_SS[[3]], by = "gene_id") %>%
  left_join(counts_SS[[4]], by = "gene_id") %>%
  left_join(counts_SS[[5]], by = "gene_id") %>%
  left_join(counts_SS[[6]], by = "gene_id") %>%
  mutate(mean_counts = rowMeans(dplyr::select(., starts_with("count")), na.rm = TRUE)) %>%
   dplyr::select(-2, -3, -4, -5, -6, -7)

mean_counts_SS
```


## Load MAJIQ TSV files

The endings RL and RD correspond to the TSV files with the different read length and read depths, respectively. 

```{r}
# load dPSI TSV files from MAJIQ Quantifier as list
dpsiDir_RL = ".../MAJIQ_TSV_files_read_length"
dpsiDir_RD = ".../MAJIQ_TSV_files_read_depth"
dpsiDir_SS = ".../MAJIQ_TSV_files_sample_size"

dpsiFiles_RL=list.files(path = dpsiDir_RL, pattern = ".tsv")
dpsiFiles_RD=list.files(path = dpsiDir_RD, pattern = ".tsv")
dpsiFiles_SS=list.files(path = dpsiDir_SS, pattern = ".tsv")

# load TSV files as data frames
setwd(dpsiDir_RL)
dPSIs_RL = lapply(dpsiFiles_RL, loadFiles)
setwd(dpsiDir_RD)
dPSIs_RD = lapply(dpsiFiles_RD, loadFiles)
setwd(dpsiDir_SS)
dPSIs_SS = lapply(dpsiFiles_SS, loadFiles)

# extract max. dPSI for each LSV and add new column with this information
dPSIs_RL = lapply(dPSIs_RL, extractMaxDPSIs)
dPSIs_RD = lapply(dPSIs_RD, extractMaxDPSIs)
dPSIs_SS = lapply(dPSIs_SS, extractMaxDPSIs)

# extract max. probability for each LSV and add a new column with this information
dPSIs_RL = lapply(dPSIs_RL, extractMaxProbs)
dPSIs_RD = lapply(dPSIs_RD, extractMaxProbs)
dPSIs_SS = lapply(dPSIs_SS, extractMaxProbs)

# only keep significant LSVs (dPSI >= 0.05, probChanging >= 0.9)
filter_function <- function(df) {
  subset(df, maxDeltaPSI >= 0.05 & maxProb >= 0.9)
}

filter_not_significant <- function(df) {
  subset(df, (maxDeltaPSI < 0.05 & maxProb < 0.9) |
             (maxDeltaPSI >= 0.05 & maxProb < 0.9) |
             (maxDeltaPSI < 0.05 & maxProb >= 0.9))
}

significant_dPSIs_RL <- lapply(dPSIs_RL, filter_function)
significant_dPSIs_RD <- lapply(dPSIs_RD, filter_function)
significant_dPSIs_SS <- lapply(dPSIs_SS, filter_function)

not_significant_dPSIs_RL <- lapply(dPSIs_RL, filter_not_significant)
not_significant_dPSIs_RD <- lapply(dPSIs_RD, filter_not_significant)
not_significant_dPSIs_SS <- lapply(dPSIs_SS, filter_not_significant)
```


# Calculate TPMs

## For U2AF2 KD experiment

```{r}
# load annotation as gtf
gtf <- rtracklayer::import(".../gencode.v41.annotation.gtf", format = "gtf")

db <- makeTxDbFromGRanges(gtf)
genes <- exonsBy(db, by = "gene")
genes <- reduce(genes)
genes2 <- lapply(genes, function(gene) sum(width(gene)))
genes3 <- unlist(genes2)
genes4 <- data.frame(gene_id = names(genes3),
                     width = genes3)

mean_counts_RL_100_2 <- merge(mean_counts_RL_100, genes4, by = "gene_id")

# divide gene counts by their genes length
mean_counts_RL_100_2$rpk <- mean_counts_RL_100_2$mean_counts / (mean_counts_RL_100_2$width/1000)

# calculate sum of all "pre-tpms" and divide raw counts by this sum and then times 1e6
scaling_factor <- sum(mean_counts_RL_100_2$mean_counts)
mean_counts_RL_100_2$tpm <- (mean_counts_RL_100_2$rpk / scaling_factor) * 1e6

mean_counts_RL_100_2
```


## For H1 cell line and endodermal cells

```{r}
mean_counts_SS_2 <- merge(mean_counts_SS, genes4, by = "gene_id")

# divide gene counts by their genes length
mean_counts_SS_2$rpk <- mean_counts_SS_2$mean_counts / (mean_counts_SS_2$width/1000)

# calculate sum of all "pre-tpms" and divide raw counts by this sum and then times 1e6
scaling_factor <- sum(mean_counts_SS_2$mean_counts)
mean_counts_SS_2$tpm <- (mean_counts_SS_2$rpk / scaling_factor) * 1e6

scaling_factor <- sum(mean_counts_SS_2$mean_counts)
mean_counts_SS_2$tpm <- (mean_counts_SS_2$rpk / scaling_factor) * 1e6

mean_counts_SS_2
```


# Fraction of TPM bins 

```{r}
# merge dPSI df with df that contains the gene counts and tpms (same read length)

# different read length
significant_dPSIs_RL[[1]] <- merge(significant_dPSIs_RL[[1]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RL[[2]] <- merge(significant_dPSIs_RL[[2]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RL[[3]] <- merge(significant_dPSIs_RL[[3]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RL[[4]] <- merge(significant_dPSIs_RL[[4]], mean_counts_RL_100_2, by = "gene_id")

# different read depth
significant_dPSIs_RD[[1]] <- merge(significant_dPSIs_RD[[1]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[2]] <- merge(significant_dPSIs_RD[[2]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[3]] <- merge(significant_dPSIs_RD[[3]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[4]] <- merge(significant_dPSIs_RD[[4]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[5]] <- merge(significant_dPSIs_RD[[5]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[6]] <- merge(significant_dPSIs_RD[[6]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[7]] <- merge(significant_dPSIs_RD[[7]], mean_counts_RL_100_2, by = "gene_id")
significant_dPSIs_RD[[8]] <- merge(significant_dPSIs_RD[[8]], mean_counts_RL_100_2, by = "gene_id")

# different sample size
significant_dPSIs_SS[[1]] <- merge(significant_dPSIs_SS[[1]], mean_counts_SS_2, by = "gene_id")
significant_dPSIs_SS[[2]] <- merge(significant_dPSIs_SS[[2]], mean_counts_SS_2, by = "gene_id")
significant_dPSIs_SS[[3]] <- merge(significant_dPSIs_SS[[3]], mean_counts_SS_2, by = "gene_id")
significant_dPSIs_SS[[4]] <- merge(significant_dPSIs_SS[[4]], mean_counts_SS_2, by = "gene_id")
```


## For each read length

```{r}
significant_dPSIs_RL[[1]] <- mutate(significant_dPSIs_RL[[1]], bin = 0)
significant_dPSIs_RL[[2]] <- mutate(significant_dPSIs_RL[[2]], bin = 0)
significant_dPSIs_RL[[3]] <- mutate(significant_dPSIs_RL[[3]], bin = 0)
significant_dPSIs_RL[[4]] <- mutate(significant_dPSIs_RL[[4]], bin = 0)

# change data type of column num_exons to make it equal in all data sets
significant_dPSIs_RL[[4]] <- mutate(significant_dPSIs_RL[[4]], num_exons = as.integer(num_exons))

all_RL <- bind_rows(
  significant_dPSIs_RL[[4]] %>% mutate(read_length = 100),
  significant_dPSIs_RL[[3]] %>% mutate(read_length = 075),
  significant_dPSIs_RL[[2]] %>% mutate(read_length = 050),
  significant_dPSIs_RL[[1]] %>% mutate(read_length = 036)
)

# bin tpm  for all data sets with different read lengths
all_RL <- all_RL %>% 
  mutate(bin = case_when(
    tpm >= 0 & tpm <= 5 ~ 1,
    tpm > 5 & tpm <= 10 ~ 2,
    tpm > 10 & tpm <= 25 ~ 3,
    tpm > 25 & tpm <= 50 ~ 4,
    tpm > 50 & tpm <= 75 ~ 5,
    tpm > 75 & tpm <= 100 ~ 6,
    tpm > 100 ~ 7,
    TRUE ~ bin
  ))

fractionTPMs_RL <- ggplot(all_RL, aes(x = factor(read_length), fill = factor(bin))) +
  geom_bar(position = position_fill(reverse = TRUE)) +
  scale_fill_manual(values = c(lighten("#FED4D4", 0.5), "#FED4D4", "#ecafaf", "#c99999", "#B28282", "#896868", darken("#896868", 0.2)),
                    labels = c("≤ 5", "≤ 10", "≤ 25", "≤ 50", "≤ 75", "<= 100", "> 100")) +
  labs(x = "read length [nt]", y = "fraction of significant LSVs [%]", title = "A", fill = "TPM") 

fractionTPMs_RL
```
```{r}
#################################
# get number of genes in each bin
#################################

# count this for 100 nt read length
all_RL_filtered <- all_RL[all_RL$read_length == 100, ]

# each gene id is only allowed to occur once
all_RL_unique_genes <- all_RL_filtered %>% distinct(gene_id, .keep_all = TRUE)

# count number of unique gene IDs for each bin
genes_per_bin <- table(all_RL_unique_genes$bin)
genes_per_bin

all_RL_filtered
all_RL_unique_genes
genes_per_bin
```


### Transform bar plot into line plot

```{r}
# aggregate data to get counts for each bin and read length
all_RL_counts <- all_RL %>%
  group_by(read_length, bin) %>%
  summarise(count = n(), .groups = 'drop')

# make line plot
fractionTPMs_RL <- ggplot(all_RL_counts, aes(x = factor(read_length), y = count, color = factor(bin), group = factor(bin))) +
  geom_line(linewidth=1.5) +
  geom_point(size=2.5) +
  scale_color_manual(values = c(
    lighten("#FED4D4", 0.5), "#FED4D4", "#ecafaf", "#c99999", "#B28282", "#896868", darken("#896868", 0.2)),
    labels = c("≤ 5", "≤ 10", "≤ 25", "≤ 50", "≤ 75", "<= 100", "> 100")
  ) +
  labs(x = "read Length [nt]", y = "number of significant LSVs", title = "A", color = "TPM")

fractionTPMs_RL
```


## For each read depth

```{r}
significant_dPSIs_RD[[1]] <- mutate(significant_dPSIs_RD[[1]], bin = 0)
significant_dPSIs_RD[[2]] <- mutate(significant_dPSIs_RD[[2]], bin = 0)
significant_dPSIs_RD[[3]] <- mutate(significant_dPSIs_RD[[3]], bin = 0)
significant_dPSIs_RD[[4]] <- mutate(significant_dPSIs_RD[[4]], bin = 0)
significant_dPSIs_RD[[5]] <- mutate(significant_dPSIs_RD[[5]], bin = 0)
significant_dPSIs_RD[[6]] <- mutate(significant_dPSIs_RD[[6]], bin = 0)
significant_dPSIs_RD[[7]] <- mutate(significant_dPSIs_RD[[7]], bin = 0)
significant_dPSIs_RD[[8]] <- mutate(significant_dPSIs_RD[[8]], bin = 0)

# change data type of column num_exons to make it equal in all data sets
significant_dPSIs_RD[[8]] <- mutate(significant_dPSIs_RD[[8]], num_exons = as.integer(num_exons))

all_RD <- bind_rows(
  significant_dPSIs_RD[[1]] %>% mutate(read_depth = 01),
  significant_dPSIs_RD[[2]] %>% mutate(read_depth = 02.5),
  significant_dPSIs_RD[[3]] %>% mutate(read_depth = 05),
  significant_dPSIs_RD[[4]] %>% mutate(read_depth = 10),
  significant_dPSIs_RD[[5]] %>% mutate(read_depth = 15),
  significant_dPSIs_RD[[6]] %>% mutate(read_depth = 20),
  significant_dPSIs_RD[[7]] %>% mutate(read_depth = 25),
  significant_dPSIs_RD[[8]] %>% mutate(read_depth = 30)
)

# bin tpm  for all data sets with different read depths
all_RD <- all_RD %>% 
  mutate(bin = case_when(
    tpm >= 0 & tpm <= 5 ~ 1,
    tpm > 5 & tpm <= 10 ~ 2,
    tpm > 10 & tpm <= 25 ~ 3,
    tpm > 25 & tpm <= 50 ~ 4,
    tpm > 50 & tpm <= 75 ~ 5,
    tpm > 75 & tpm <= 100 ~ 6,
    tpm > 100 ~ 7,
    TRUE ~ bin
  ))

fractionTPMs_RD <- ggplot(all_RD, aes(x = factor(read_depth), fill = factor(bin))) +
  geom_bar(position = position_fill(reverse = TRUE)) +
  scale_fill_manual(values = c(lighten("#FED4D4", 0.5), "#FED4D4", "#ecafaf", "#c99999", "#B28282", "#896868", darken("#896868", 0.2)),
                    labels = c("≤ 5", "≤ 10", "≤ 25", "≤ 50", "≤ 75", "<= 100", "> 100")) +
  labs(x = "read depth [10^6]", y = "fraction of significant LSVs [%]", title = "B", fill = "TPM") 

fractionTPMs_RD
```

### Transform bar plot into line plot
```{r}
# aggregate data to get counts for each bin and read depth
all_RD_counts <- all_RD %>%
  group_by(read_depth, bin) %>%
  summarise(count = n(), .groups = 'drop')

# make line plot
fractionTPMs_RD <- ggplot(all_RD_counts, aes(x = factor(read_depth), y = count, color = factor(bin), group = factor(bin))) +
  geom_line(linewidth=1.5) +
  geom_point(size=2.5) +
  scale_color_manual(values = c(
    lighten("#FED4D4", 0.5), "#FED4D4", "#ecafaf", "#c99999", "#B28282", "#896868", darken("#896868", 0.2)),
    labels = c("≤ 5", "≤ 10", "≤ 25", "≤ 50", "≤ 75", "<= 100", "> 100")
  ) +
  labs(x = "read depth [10^6]", y = "number of significant LSVs", title = "A", color = "TPM")

fractionTPMs_RD
```


## For different numbers of replicates

```{r}
significant_dPSIs_SS[[1]] <- mutate(significant_dPSIs_SS[[1]], bin = 0)
significant_dPSIs_SS[[2]] <- mutate(significant_dPSIs_SS[[2]], bin = 0)
significant_dPSIs_SS[[3]] <- mutate(significant_dPSIs_SS[[3]], bin = 0)
significant_dPSIs_SS[[4]] <- mutate(significant_dPSIs_SS[[4]], bin = 0)

# different sample combinations named the same here, as a stacked bar chart normalised to 1 is made 
# the ratios therefore remain the same
all_SS <- bind_rows(
  significant_dPSIs_SS[[4]] %>% mutate(samples = 3),
  significant_dPSIs_SS[[3]] %>% mutate(samples = 2),
  significant_dPSIs_SS[[2]] %>% mutate(samples = 2),
  significant_dPSIs_SS[[1]] %>% mutate(samples = 2)
)

# bin tpms for all data sets with different replicate numbers
all_SS <- all_SS %>% 
  mutate(bin = case_when(
    tpm >= 0 & tpm <= 5 ~ 1,
    tpm > 5 & tpm <= 10 ~ 2,
    tpm > 10 & tpm <= 25 ~ 3,
    tpm > 25 & tpm <= 50 ~ 4,
    tpm > 50 & tpm <= 75 ~ 5,
    tpm > 75 & tpm <= 100 ~ 6,
    tpm > 100 ~ 7,
    TRUE ~ bin
  ))

fractionTPMs_SS <- ggplot(all_SS, aes(x = factor(samples), fill = factor(bin))) +
  geom_bar(position = position_fill(reverse = TRUE)) +
  scale_fill_manual(values = c(lighten("#FED4D4", 0.5), "#FED4D4", "#ecafaf", "#c99999", "#B28282", "#896868", darken("#896868", 0.2)),
                    labels = c("≤ 5", "≤ 10", "≤ 25", "≤ 50", "≤ 75", "<= 100", "> 100")) +
  labs(x = "# replicates", y = "fraction of significant LSVs [%]", title = "C", fill = "TPM") 

fractionTPMs_SS
```


### Transform bar plot into line plot

```{r}
# aggregate data to get counts for each bin and number of replicates
all_SS_counts <- all_SS %>%
  group_by(samples, bin) %>%
  summarise(count = n(), .groups = 'drop')

all_SS_counts

# make line plot
fractionTPMs_SS <- ggplot(all_SS_counts, aes(x = factor(samples), y = count, color = factor(bin), group = factor(bin))) +
  geom_line(linewidth=1.5) +
  geom_point(size=2.5) +
  scale_color_manual(values = c(
    lighten("#FED4D4", 0.5), "#FED4D4", "#ecafaf", "#c99999", "#B28282", "#896868", darken("#896868", 0.2)),
    labels = c("≤ 5", "≤ 10", "≤ 25", "≤ 50", "≤ 75", "<= 100", "> 100")
  ) +
  labs(x = "# replicates", y = "number of significant LSVs", title = "A", color = "TPM")

fractionTPMs_SS
```
