# Impact of RNASeq read depth and read length on alternative splicing analysis
****
This Github repository provides the code the for the publication Ladwig et al. 2023.

## 1 Preprocessing and alternative splicing analysis with MAJIQ
****
The RMarkdown contains the preprocessing steps of the raw reads and the following MAJIQ scripts for alternative splicing analysis. 

benötigt für das preprocessing: conda, ...

The rendered HTML-report can be accessed at ... .



## 2 Impact of library characteristics on alternative splice analysis
****
The R scripts contain the code used to investigate the influence of read depth, read length and sample size on: 
- read alignment
- LSV detection
- LSV quantification
- alternative splicing event types distribution



